lib/streamlit/elements/widgets/data_editor.py

```diff
@@ -817,15 +817,36 @@ def data_editor(
 
         # Convert the user provided column config into the frontend compatible format:
         column_config_mapping = process_config_mapping(column_config)
-        apply_data_specific_configs(
-            column_config_mapping, data_df, data_format, check_arrow_compatibility=True
-        )
+
+        # Deactivate editing for columns that are not compatible with arrow
+        for column_name, column_data in data_df.items():
+            if dataframe_util.is_colum_type_arrow_incompatible(column_data):
+                update_column_config(
+                    column_config_mapping, column_name, {"disabled": True}
+                )
+                # Convert incompatible type to string
+                data_df[column_name] = column_data.astype("string")
+
+        apply_data_specific_configs(column_config_mapping, data_format)
 
         # Fix the column headers to work correctly for data editing:
         _fix_column_headers(data_df)
-        # Temporary workaround: We hide range indices if num_rows is dynamic.
-        # since the current way of handling this index during editing is a bit confusing.
-        if isinstance(data_df.index, pd.RangeIndex) and num_rows == "dynamic":
+
+        has_range_index = isinstance(data_df.index, pd.RangeIndex)
+
+        if not has_range_index:
+            # If the index is not a range index, we will configure it as required
+            # since the user is required to provide a (unique) value for editing.
+            update_column_config(
+                column_config_mapping, INDEX_IDENTIFIER, {"required": True}
+            )
+
+        if has_range_index and num_rows == "dynamic":
```

Comment:
I think this can stay an `elif` as the `not has_range_index` and `has_range_index` are exclusive. 
I guess the more confusing point is that you set it here but then it can be overridden in [line 854](https://github.com/streamlit/streamlit/pull/9205/files#diff-e591171c22c1c88a8244dbf9fd17e4f30b2d9d9cd93f8bac5479cd1ecd992448R854), which makes it be executed twice.

I think we could merge it like this:

```python
_hide_index = hide_index
if _hide_index is None and has_range_index and num_rows == "dynamic":
  _hide_index = True

if _hide_index is not None:
  update_column_config(
      column_config_mapping, INDEX_IDENTIFIER, {"hidden": hide_index}
  )
```

but I acknowledge that the check might also look confusing 🤔

Reply:
Update the code 👍